libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/StepFunctionHelpers.java

```diff
@@ -53,16 +56,33 @@ public static List<String> getFeatureSet(Map<String, String> input) {
     }
 
     public static String getJourneyEvent(Map<String, String> input)
+            throws HttpResponseExceptionWithErrorBody, UnsupportedEncodingException {
+        String journeyEvent =
+                Optional.ofNullable(input.get(JOURNEY))
+                        .orElseThrow(
+                                () ->
+                                        new HttpResponseExceptionWithErrorBody(
+                                                HttpStatus.SC_BAD_REQUEST,
+                                                ErrorResponse.MISSING_JOURNEY_EVENT));
+
+        String[] decodedParts = URLDecoder.decode(journeyEvent, StandardCharsets.UTF_8).split("/");
+
+        return decodedParts[decodedParts.length - 1];
+    }
+
+    public static Optional<String> getCurrentPage(Map<String, String> input)
             throws HttpResponseExceptionWithErrorBody {
-        String[] parts =
+        String journeyEvent =
                 Optional.ofNullable(input.get(JOURNEY))
                         .orElseThrow(
                                 () ->
                                         new HttpResponseExceptionWithErrorBody(
                                                 HttpStatus.SC_BAD_REQUEST,
-                                                ErrorResponse.MISSING_JOURNEY_EVENT))
-                        .split("/");
-        return parts[parts.length - 1];
+                                                ErrorResponse.MISSING_JOURNEY_EVENT));
+
+        String[] decodedParts = URLDecoder.decode(journeyEvent, StandardCharsets.UTF_8).split("/");
+
+        return Optional.ofNullable(decodedParts.length == 4 ? decodedParts[2] : null);
```

Comment:
What if there are more parts? e.g. if I maliciously submit an event for `foo/next`, can I bypass the validation?

(This is the sort of reason why supplying the page id as a separate parameter would be nicer 😬)

Reply:
Actually we would need to permanently support two separate API Gateways, one for all user-facing journey-event requests, and one for internal journey-event requests.

(Unless we changed all the internal journey-event requests to contain a pageId as well, which would involve updating the step function I believe)

Reply:
Yeah that's true, we might need to name the gateway address something different from `/journey/*` so we can differentiate between them easily in the meantime.

I think that would be my preferred approach, if we are going to this length to refactor and make the journey more robust, we might as well do it properly